### ISSUE TITLE
[BUG] mocktracer start is not threadsafe, although setting the globaltracer is threadsafe

### ISSUE THREAD

**Issue by valsteen** [DataDog/dd-trace-go#1700]
The global `Testing` flag in `ddtrace/internal/globaltracer.go` is not thread-safe, which can lead to a datarace crashes when running tests with `-race`. I would tend to believe it was overlooked, since `SetGlobalTracer` is itself threadsafe ( `globalTracer` is an `atomic.Value` ).

I have tests scenarios starting a webserver in a goroutine. That webserver calls `tracer.Start` which checks for the `Testing` flag.

https://github.com/DataDog/dd-trace-go/blob/1160be5c0b9705d183fbd80c37ea73add96af669/ddtrace/tracer/tracer.go#L121-L124

While the server starts, my test setup starts the `mocktracer`. It sets `internal.Testing`.

https://github.com/DataDog/dd-trace-go/blob/1160be5c0b9705d183fbd80c37ea73add96af669/ddtrace/mocktracer/mocktracer.go#L50-L55

Currently I have to use a workaround - unfortunately I can't directly control when `tracer.Start` is called, I would have to either fork the web framework I'm using or ddtrace to solve this issue properly

**Version of dd-trace-go**

`v1.46.1`. As I can see the issue still exist on `main` branch.

**Describe what happened:**

This can be reproduced by this snippet that is run with `-race`:

```go
	go func() {
		tracer.Start()
	}()

	go func() {
		mocktracer.Start()
	}()
```

```
WARNING: DATA RACE
Write at 0x000104cc282c by goroutine 23:
  gopkg.in/DataDog/dd-trace-go.v1/ddtrace/mocktracer.Start()
      .../pkg/mod/gopkg.in/!data!dog/dd-trace-go.v1@v1.46.0/ddtrace/mocktracer/mocktracer.go:53 +0xd4

Previous read at 0x000104cc282c by goroutine 22:
  gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer.Start()
      .../pkg/mod/gopkg.in/!data!dog/dd-trace-go.v1@v1.46.0/ddtrace/tracer/tracer.go:122 +0x3c
```

**Describe what you expected:**

tracer.Start() and mocktracer.Start() are safe to run concurrently

**Steps to reproduce the issue:**

```go
	go func() {
		tracer.Start()
	}()

	go func() {
		mocktracer.Start()
	}()
```

**Additional environment details (Version of Go, Operating System, etc.):**

tests are run in a `golang:1.19.0-buster` docker image. Latest versions give the same result.

**Comment by knusbaum** [comment]
I have a fix up that makes `tracer.Start()` and `mocktracer.Start()` safe to run concurrently.

Keep in mind that this just makes those operations atomic. Exactly one of `tracer.Start()` and `mocktracer.Start()` will "win" and their tracer will be set correctly. However, which one will win is down to timing, and needs to be synchronized to ensure which one you will end up with.

That is not a problem we can solve within the library.

**Comment by valsteen** [comment]
@knusbaum yes, I'm aware of that, it may not work as expected for all users - in my case I directly use the tracer instance, not the global one, so this is fine
